### PR TITLE
feat: update sort algorithm using loser tree for multi sort merge

### LINKED ISSUE #11604: Feature: Update sort algorithm using Loser Tree
**Summary**

Currently Databend's sort algorithm using MergSort based on **heap**. The main algorithms of multiple merging include heap sort, winner tree and loser tree. In these three algorithms, each heap adjustment of heap sorting requires comparison with the left and right child nodes, and the comparison times is 2logN, while the comparison times of winner tree and loser tree adjustment are logN. The difference is that the winner tree needs to compare with sibling nodes and update the parent node, while the loser tree only needs to compare with the parent node, and the memory access times are less.

### PR DESCRIPTION
**Pull request by forsaken628**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Fixes #11604

## Tests

- [x] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15869)

### CODE REVIEW

**Comment by sundy-li** [comment]
Is there any performance comparison data of this pr?

**Comment by forsaken628** [comment]
> Is there any performance comparison data of this pr?

Not yet, do I need to add an algorithm level benchmark, or just ci-benchmark?

**Comment by sundy-li** [comment]
you can pref it in local, eg:

```
create table t(a int, b string, d float, e date);
create table t_random like t engine = random;

## generate data
insert into t select * from t_random limit 5000000;
insert into t select * from t_random limit 5000000;
insert into t select * from t_random limit 5000000;
insert into t select * from t_random limit 5000000;
...

## sort perf 

select * from t order by a,b ignore_result;
...
```

**Comment by forsaken628** [comment]
> you can pref it in local, eg:
> 
> ```
> create table t(a int, b string, d float, e date);
> create table t_random like t engine = random;
> 
> ## generate data
> insert into t select * from t_random limit 5000000;
> insert into t select * from t_random limit 5000000;
> insert into t select * from t_random limit 5000000;
> insert into t select * from t_random limit 5000000;
> ...
> 
> ## sort perf 
> 
> select * from t order by a,b ignore_result;
> ...
> ```

No noticeable changes. There seems to be a slight improvement from flamegraph, not obvious. It could also be due to the distribution of the data.

flamegraph

## data set

read rows: 25000000
read size: 202.54 MiB
partitions total: 24
partitions scanned: 24
pruning stats: [segments: <range pruning: 1 to 1>, blocks: <range pruning: 24 to 24>]
push downs: [filters: [], limit: NONE]
estimated rows: 25000000.00

## sql

### sort
select * from t order by a,b ignore_result;
[pprof.cpu.heap.sort.pb.gz](https://github.com/user-attachments/files/15981752/pprof.cpu.heap.sort.pb.gz)
[pprof.cpu.loser.sort.pb.gz](https://github.com/user-attachments/files/15981754/pprof.cpu.loser.sort.pb.gz)

### window 
SELECT MAX(d) OVER (PARTITION BY a) FROM t ignore_result;
[pprof.cpu.heap.window.pb.gz](https://github.com/user-attachments/files/15981753/pprof.cpu.heap.window.pb.gz)
[pprof.cpu.loser.window.pb.gz](https://github.com/user-attachments/files/15981755/pprof.cpu.loser.window.pb.gz)

**Comment by sundy-li** [comment]
The codes LGTM. Some comments about this pr:

Better add a new setting named `enable_loser_tree_merge_sort`, default to `1`.

Then we can create the `merger` in `MultiSortMergeProcessor` by setting in runtime. If any bug happens, we can switch to use different impl.

File: src/query/pipeline/transforms/src/processors/transforms/sort/rows/simple.rs [hunk 1]
```diff
@@ -81,7 +81,7 @@ where
     T::Scalar: Ord,
 {
     fn eq(&self, other: &Self) -> bool {
-        self == other
+        self.inner == other.inner && self.desc == other.desc
```
Comment (forsaken628) [inline src/query/pipeline/transforms/src/processors/transforms/sort/rows/simple.rs:84]: Should desc be ignored or is it inconsistent with cmp behavior.

Comment (sundy-li) [reply src/query/pipeline/transforms/src/processors/transforms/sort/rows/simple.rs:84]: the descs are consistent.

File: src/query/settings/src/settings_default.rs [hunk 2]
```diff
@@ -775,6 +775,12 @@ impl DefaultSettings {
                     desc: "The maximum count of set operator in a query. If your query stack overflow, you can reduce this value.",
                     mode: SettingMode::Both,
                     range: Some(SettingRange::Numeric(0..=u64::MAX)),
+                }),
+                ("enable_loser_tree_merge_sort", DefaultSettingValue {
+                    value: UserSettingValue::UInt64(1),
```
Comment (sundy-li) [inline src/query/settings/src/settings_default.rs:780]: I still did not find `loser_tree` is faster than `heap`.

```
🐳 :) select l_orderkey from lineitem order by l_orderkey ignore_result;
0 row read in 1.409 sec. Processed 59.99 million row, 457.66 MiB (42.58 million rows/s, 324.82 MiB/s)

🐳 :) set enable_loser_tree_merge_sort = 0;
🐳 :) select l_orderkey from lineitem order by l_orderkey ignore_result;
0 row read in 1.258 sec. Processed 59.99 million row, 457.66 MiB (47.7 million rows/s, 363.90 MiB/s)
```

```

🐳 :) set enable_loser_tree_merge_sort = 1;
🐳 :) select userid from hits order by userid ignore_result;
0 row read in 3.027 sec. Processed 100 million row, 762.92 MiB (33.04 million rows/s, 252.08 MiB/s)
🐳 :) set enable_loser_tree_merge_sort = 0;
🐳 :) select userid from hits order by userid ignore_result;
0 row read in 2.528 sec. Processed 100 million row, 762.92 MiB (39.56 million rows/s, 301.83 MiB/s)
```

Comment (sundy-li) [reply src/query/settings/src/settings_default.rs:780]: Maybe we should disable it by default unless we find it's faster.

But the sort algorithm abstraction is fine.

Comment (sundy-li) [reply src/query/settings/src/settings_default.rs:780]: Is it true that when K is large enough, the performance of the loser tree will be better?

Current K is the num of physical cores by default.

Comment (forsaken628) [reply src/query/settings/src/settings_default.rs:780]: https://github.com/datafuselabs/databend/blob/7b15120fd2b2f91cf3a3f9ca860ea733f830eefc/src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs#L171
I think the problem is with this peek_top2.
It is generally accepted that the loser tree is twice as fast as the heap, but the need to fetch top2 here causes the loser tree to be accessed twice per iteration. And the standard library's heap is deeply optimized. Maybe some kind of variant implementation can be found later to improve this situation.

**Review by sundy-li — APPROVED** [review]

File: src/query/settings/src/settings_default.rs (hunk 2 above)
Comment (sundy-li) [reply src/query/settings/src/settings_default.rs]: Tracked in https://github.com/datafuselabs/databend/issues/15925

**Comment by forsaken628** [comment]
Here's another part about Merger's changes.

https://github.com/forsaken628/databend/compare/loser-tree...forsaken628:databend:peek-mut

The cursor contains Arc, and iterative cloning and deletion incurs a lot of overhead, so I'm replacing cloning with in-place updating.

Should these two parts of the changes be separately PR or combined?

**Comment by sundy-li** [comment]
> Should these two parts of the changes be separately PR or combined?

You can add into this pr directly.

**Review by zhang2014 — APPROVED** [review]

**Pull request merged**